spawn_worker/src/unix/mod.rs

```diff
@@ -45,53 +54,56 @@ pub unsafe fn get_dl_path_raw(addr: FnPointer) -> (Option<CString>, Option<CStri
     (path_name, symbol_name)
 }
 
-pub enum Fork {
-    Parent(libc::pid_t),
-    Child,
+/// Returns PID of current process
+pub fn getpid() -> libc::pid_t {
+    unsafe { libc::getpid() }
 }
 
-/// Forks process into a new standalone process
-///
-/// # Errors
-///
-/// This function will return an error if child process can't be forked
-///
-/// # Safety
-///
-/// Existing state of the process must allow safe forking, e.g. no background threads should be running
-/// as any locks held by these threads will be locked forever
-///
-/// When forking a multithreaded application, no code should allocate or access other potentially locked resources
-/// until call to exec is executed
-pub unsafe fn fork() -> Result<Fork, std::io::Error> {
-    let res = libc::fork();
-    match res {
-        -1 => Err(std::io::Error::last_os_error()),
-        0 => Ok(Fork::Child),
-        res => Ok(Fork::Parent(res)),
+impl From<Entrypoint> for spawn::Target {
+    fn from(entrypoint: Entrypoint) -> Self {
+        spawn::Target::Entrypoint(entrypoint)
     }
 }
 
-/// Runs supplied closure in separate process via fork(2)
-///
-/// # Safety
-///
-/// Existing state of the process must allow safe forking, e.g. no background threads should be running
-/// as any locks held by these threads will be locked forever
-///
-/// When forking a multithreaded application, no code should allocate or access other potentially locked resources
-/// until call to exec is executed
-pub unsafe fn fork_fn<Args>(args: Args, f: fn(Args) -> ()) -> Result<libc::pid_t, std::io::Error> {
-    match fork()? {
-        Fork::Parent(pid) => Ok(pid),
-        Fork::Child => {
-            f(args);
-            std::process::exit(0)
-        }
+impl Entrypoint {
+    pub fn get_fs_path(&self) -> Option<PathBuf> {
+        let (path, _) = unsafe { get_dl_path_raw(self.ptr as *const libc::c_void) };
+
+        Some(PathBuf::from(path?.to_str().ok()?.to_owned()))
     }
 }
 
-/// Returns PID of current process
-pub fn getpid() -> libc::pid_t {
-    unsafe { libc::getpid() }
+pub(crate) static ENV_PASS_FD_KEY: &str = "__DD_INTERNAL_PASSED_FD";
+
+pub fn recv_passed_fd() -> Option<OwnedFd> {
+    let val = env::var(ENV_PASS_FD_KEY).ok()?;
+    let fd: RawFd = val.parse().ok()?;
+
+    // check if FD number is valid
+    nix::fcntl::fcntl(fd, nix::fcntl::FcntlArg::F_GETFD).ok()?;
+
+    Some(unsafe { OwnedFd::from_raw_fd(fd) })
+}
```

Comment:
This reminds me that we should probably close all non passed file descriptors, in order to not keep alive the applications files and socket open, since the sidecar could outlive the parent process

Reply:
That is the beauty of using exec! we don't have to! 
Assuming we can trust most cases where people might have some sockets, and if they don't like those sockets to be shared with other processes they will do set `CLOSEXEC`

....

Now I realized that its the APM world we're talking about here :| - and maybe we should close them